File: joss/comparisons/cat-6-contextual-runtime/09-forbidden-methods/example-forbidden-methods.py
"""
Error Class 9: Unintended Escape from DataFrame API

Python DataFrames don't prevent you from using raw Python operations
that bypass the DataFrame API. This can silently produce wrong results
or lose DataFrame metadata.
"""
import pandas as pd

patients = pd.DataFrame({
    "patient_id": ["P001", "P002", "P003"],
    "first_name": ["Maria", "James", "Abigail"],
    "age": [66, 49, 34],
})

# ── PROBLEM 9a: Using list comprehension instead of vectorized ops ──────
# Works but returns a plain list — loses DataFrame structure and index.
# No warning that you've left the pandas world.
names = [row["first_name"] for _, row in patients.iterrows()]
# Result is a plain Python list, not a Series or DataFrame

# ── PROBLEM 9b: Using .apply() with side effects ───────────────────────
# .apply() accepts any callable — no type checking on return values.
# If the function returns inconsistent types, pandas silently coerces.
def process(row):
    if row["age"] > 50:
        return row["age"]  # returns int
    return "young"          # returns str — column becomes object type silently

patients["status"] = patients.apply(process, axis=1)
print(patients["status"].dtype)  # object — mixed types, no warning

# ── PROBLEM 9c: Direct mutation via assignment ─────────────────────────
# pandas allows in-place mutation. No immutability guarantees.
patients.loc[0, "age"] = -5  # Invalid age — no type or range check

File: joss/comparisons/cat-6-contextual-runtime/09-forbidden-methods/example-forbidden-methods.R
# Error Class 9: Unintended Escape from DataFrame API
#
# R tibbles don't prevent you from using base R operations that
# bypass tidyverse safety. Direct indexing, $ access, and base
# functions can silently produce wrong results.

library(tidyverse)

patients <- tibble(
  patient_id = c("P001", "P002", "P003"),
  first_name = c("Maria", "James", "Abigail"),
  age = c(66, 49, 34)
)

# ── PROBLEM 9a: Using base R $ with typo ───────────────────────────────
# SILENT: $ on tibble with wrong column name returns NULL — no error.
names <- patients$fist_name  # Typo: 'fist_name' instead of 'first_name'
print(names)
# NULL — no error, no warning. Downstream code using this will fail
# or produce wrong results much later.

# ── PROBLEM 9b: Direct vector assignment bypasses type safety ───────────
# No type checking on assignment. You can put anything in any column.
patients$age[1] <- "old"  # Coerces entire column to character — silently
print(patients$age)
# [1] "old" "49"  "34" — all ages are now strings

# ── PROBLEM 9c: sapply returns inconsistent types ──────────────────────
# sapply() can return a vector, matrix, or list depending on results.
# No type guarantee on what comes back.
result <- sapply(patients$age, function(x) {
  if (x > 50) return(x)
  return("young")
})
# Returns a character vector — coerced the numbers to strings silently
